### PR TITLE
Add S3 bucket and IAM Role/Policy for user NFS home dir backups

### PR DESCRIPTION
**Pull request by kerin**
## What

Added:
- S3 bucket with versioning and encryption enabled
- IAM role for backup container
- IAM policy for container access to S3 bucket

## How to review

1. Checkout this branch
2. `terraform plan` and `terraform apply` in dev env
3. Verify that bucket, IAM role and IAM policy are created
4. Install the Helm chart from [this related PR](https://github.com/ministryofjustice/analytics-platform-helm-charts/pull/33) and confirm backups are successful

### CODE REVIEW

File: infra/terraform/environments/dev/main.tf [hunk 1]
```diff
@@ -59,6 +59,13 @@ module "user_nfs_softnas" {
     dns_zone_domain = "${module.cluster_dns.dns_zone_domain}"
 }
 
+module "data_backup" {
+    source = "../../modules/data_backup"
+
+    env = "${var.env}"
+    k8s_worker_role_arn = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/nodes.${var.env}.${data.terraform_remote_state.base.xyz_root_domain}"
```
Comment (xoen) [inline infra/terraform/environments/dev/main.tf]: Would be nice to have interpolation in the `.tfvars` or intermediate variables...but currently it's not possible ([TF #4084](https://github.com/hashicorp/terraform/issues/4084) and [TF #10059](https://github.com/hashicorp/terraform/issues/10059)).

File: infra/terraform/modules/data_backup/data.tf [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      type        = "AWS"
+      identifiers = ["${var.k8s_worker_role_arn}"]
+    }
+  }
+}
+
+data "aws_iam_policy_document" "nfs_backup" {
+  statement {
+    effect = "Allow"
+    actions = [
```
Comment (xoen) [inline infra/terraform/modules/data_backup/data.tf]: This looks good. I still need to have a look at the other PR, just saw it's using `s3cmd`, I don't know exactly what kind of operations will try to perform.

I was wondering if:
 * Will it need to perform other multipart-related actions?
 * Does it need to operate on versions?

It's good to start by keeping the policies to the minimum as in this case.

File: infra/terraform/modules/data_backup/data.tf [hunk 3]
```diff
@@ -0,0 +1,51 @@
+data "aws_region" "current" {
+  current = true
+}
+
+data "aws_iam_policy_document" "assume_role_policy" {
```
Comment (xoen) [inline infra/terraform/modules/data_backup/data.tf]: Minor style point. I would probably prefer to have these IAM-related resources near the others using them (the `iam.tf` file).

It obviously doesn't make any difference from Terraform perspective - so don't worry about this.

File: infra/terraform/modules/data_backup/data.tf (hunk 2 above)
Comment (kerin) [reply infra/terraform/modules/data_backup/data.tf]: So far in my testing it hasn't needed any other permissions - other multipartupload actions would be required for large file uploads, but >100MB uploads have worked successfully - similarly s3cmd doesn't interact with versions directly and just uploads/deletes files, with the bucket itself taking care of versioning.

File: infra/terraform/modules/data_backup/data.tf (hunk 3 above)
Comment (kerin) [reply infra/terraform/modules/data_backup/data.tf]: yep, fair enough - I've moved the IAM documents into `iam.tf`

**Pull request merged**